### ISSUE TITLE
Type inference of Tuples returns object instead

### ISSUE THREAD

**Issue by eddieschoute** [python/mypy#4975]
While interacting with `zip` I encountered a variant of the following error:
```
error: No overload variant of "zip" matches argument types [builtins.list[builtins.int*], builtins.object*]
```
which is emitted by mypy on the case of
```py
out = list(
    list(zip([1, 2], ordered_item))
    for item in [(1, 2), (4, 5)]
    for ordered_item in [item, tuple(reversed(item))]
    )
print(out)
```
This code runs normally on python 3.6.5, outputting:
```
[[(1, 1), (2, 2)], [(1, 2), (2, 1)], [(1, 4), (2, 5)], [(1, 5), (2, 4)]]
```

## Expected Behavior
Mypy should infer that `ordered_item` is a tuple, or at least an iterable so that `zip` can be applied on it.

It may also be interesting to consider the case where `ordered_item` is heterogenous over different `Iterables` (i.e. not only tuples). Would Iterable be inferred then?

## Actual Behavior
Mypy infers the `object` type on `ordered_item` which you understandably cannot zip.

By explicitly casting `item` to tuple the error message is resolved:
```py
out = list(
    list(zip([1, 2], ordered_item))
    for item in [(1, 2), (4, 5)]
    for ordered_item in [tuple(item), tuple(reversed(item))]
    )
print(out)
```

## Environment
OS: macOS 10.13
Python: 3.6.5
mypy: 0.590

**Comment by beamerblvd** [comment]
Is this fix available in a release?

**Comment by ethanhs** [comment]
@beamerblvd I'm not sure I understand. I don't think there is a fix for this yet

**Comment by beamerblvd** [comment]
@ethanhs, my mistake. I saw a red `Closed` and thought it was for this issue, but it was for another issue having reference this.

**Comment by ilevkivskyi** [comment]
This already appeared five times, so I am raising priority to high (especially since IIUC we are not going to infer unions for ternary expressions). This should be not hard to fix if someone wants to try.

**Comment by intgr** [comment]
Another minimal reproducer:
```python
tup1: tuple = ()
tup2 = tup1 if input() else ()
reveal_type(tup2)   # N: Revealed type is 'builtins.object'
```
What is even happening here? 😄

**Comment by JelleZijlstra** [comment]
@intgr the underlying problem is that internally in mypy, the types for fixed-length tuples (like `Tuple[int, int]` and variadic tuples (`Tuple[int, ...]`) are completely separate, so when mypy tries to join these two types, it lands on `object` instead of something more reasonable.

**Comment by JukkaL** [comment]
@intgr Do you think that this issue can be closed now, after you've PRs were merged?

**Comment by intgr** [comment]
Yes, that should solve all the cases involving tuples. Also #8074 can be closed, as it falls back to `Sequence[]` type.

**Comment by ethanhs** [comment]
Closing as it was fixed in #8335. I'm very happy to see this issue fixed :)
